lib/model/model.go

```diff
@@ -252,24 +254,41 @@ func NewModel(cfg config.Wrapper, id protocol.DeviceID, clientName, clientVersio
 }
 
 func (m *model) serve(ctx context.Context) error {
-	// Add and start folders
+	m.initFolders()
+	m.cfg.Subscribe(m)
+	close(m.started)
+
+	var err error
+	select {
+	case <-ctx.Done():
+	case err = <-m.fatalChan:
+	}
+
+	m.cfg.Unsubscribe(m)
+	m.closeAllConnectionsAndWait()
+
+	return err
+}
+
+func (m *model) initFolders() {
 	cacheIgnoredFiles := m.cfg.Options().CacheIgnoredFiles
 	clusterConfigDevices := make(deviceIDSet, len(m.cfg.Devices()))
 	for _, folderCfg := range m.cfg.Folders() {
 		if folderCfg.Paused {
 			folderCfg.CreateRoot()
 			continue
 		}
-		m.newFolder(folderCfg, cacheIgnoredFiles)
+		err := m.newFolder(folderCfg, cacheIgnoredFiles)
+		if err != nil {
+			m.fatal(err)
```

Comment:
This will deadlock, because fatalChan is unbuffered and we only start reading from it after initFolders returns? Perhaps have initFolders return err instead and let the caller handle it?